From 2c4ae87140cc719a7e94a59b421374bbeee898c2 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Olivier=20Cr=C3=AAte?= <olivier.crete@collabora.com>
Date: Mon, 5 Nov 2018 14:57:21 -0500
Subject: [PATCH 1/3] scanner: Pass library paths before the -l itself

Without this, it will probably take the system library instead of the
one that we are trying to test.
---
 giscanner/ccompiler.py | 22 +++++++++++-----------
 1 file changed, 11 insertions(+), 11 deletions(-)

diff --git a/giscanner/ccompiler.py b/giscanner/ccompiler.py
index c0038285..78e3825f 100644
--- a/giscanner/ccompiler.py
+++ b/giscanner/ccompiler.py
@@ -133,17 +133,6 @@ class CCompiler(object):
                 if sys.platform != 'darwin':
                     args.append('-Wl,--no-as-needed')
 
-        for library in libraries + extra_libraries:
-            if self.check_is_msvc():
-                # Note that Visual Studio builds do not use libtool!
-                if library != 'm':
-                    args.append(library + '.lib')
-            else:
-                if library.endswith(".la"):  # explicitly specified libtool library
-                    args.append(library)
-                else:
-                    args.append('-l' + library)
-
         for library_path in libpaths:
             # The dumper program needs to look for dynamic libraries
             # in the library paths first
@@ -161,6 +150,17 @@ class CCompiler(object):
 
             runtime_paths.append(library_path)
 
+        for library in libraries + extra_libraries:
+            if self.check_is_msvc():
+                # Note that Visual Studio builds do not use libtool!
+                if library != 'm':
+                    args.append(library + '.lib')
+            else:
+                if library.endswith(".la"):  # explicitly specified libtool library
+                    args.append(library)
+                else:
+                    args.append('-l' + library)
+
         for envvar in runtime_path_envvar:
             if envvar in os.environ:
                 os.environ[envvar] = \
-- 
2.18.1


From a4cfd0a8e5542e333810bd94091a5176cbc2901a Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Olivier=20Cr=C3=AAte?= <olivier.crete@collabora.com>
Date: Mon, 5 Nov 2018 15:41:12 -0500
Subject: [PATCH 2/3] scanner: Accept library filenames

This makes it easier to ensure that the right library is
scanned.
---
 giscanner/ccompiler.py | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/giscanner/ccompiler.py b/giscanner/ccompiler.py
index 78e3825f..d327c581 100644
--- a/giscanner/ccompiler.py
+++ b/giscanner/ccompiler.py
@@ -156,7 +156,8 @@ class CCompiler(object):
                 if library != 'm':
                     args.append(library + '.lib')
             else:
-                if library.endswith(".la"):  # explicitly specified libtool library
+                # If we get a real filename, just use it as-is
+                if library.endswith(".la") or os.path.isfile(library):
                     args.append(library)
                 else:
                     args.append('-l' + library)
-- 
2.18.1


From 630f6626a89322f62b0709c3bab686c4c555c73c Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Olivier=20Cr=C3=AAte?= <olivier.crete@collabora.com>
Date: Mon, 5 Nov 2018 16:57:57 -0500
Subject: [PATCH 3/3] scanner: Skip ldd for library by filename

---
 giscanner/shlibs.py | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/giscanner/shlibs.py b/giscanner/shlibs.py
index 7b7b2d02..f70c54da 100644
--- a/giscanner/shlibs.py
+++ b/giscanner/shlibs.py
@@ -119,7 +119,10 @@ def _resolve_non_libtool(options, binary, libraries):
         proc = subprocess.Popen(args, stdout=subprocess.PIPE)
         patterns = {}
         for library in libraries:
-            patterns[library] = _ldd_library_pattern(library)
+            if not os.path.isfile(library):
+                patterns[library] = _ldd_library_pattern(library)
+        if len(patterns) == 0:
+            return []
 
         shlibs = []
         for line in proc.stdout:
-- 
2.18.1

